pkg/kubectl/cmd/auth/cani.go

```diff
@@ -40,7 +40,7 @@ type CanIOptions struct {
 	AllNamespaces bool
 	Quiet         bool
 	Namespace     string
-	SelfSARClient internalauthorizationclient.SelfSubjectAccessReviewsGetter
+	SelfSARClient authorizationv1client.AuthorizationV1Interface
```

Comment:
This pattern is going to cause skew problems in the future.  Do you have a plan for dealing with it?

Reply:
What skew problems are you worried about?  The old API will continue to work right?

Reply:
@lavalamp I believe this is what you suggested we do instead of using the internal types right?

Reply:
> What skew problems are you worried about? The old API will continue to work right?

It means that when alpha is replaced by beta, commands like this would have to choose whether they want to work with the old servers or the new servers or duplicate the code.  The same issue can come up with any version transition.

Reply:
They need to decide if they want to work with the old API or the new API right?  There should be some period of time in which both APIs are supported by the server.  How is this different from what clients built outside kubernetes/kubernetes need to do?

Reply:
@deads2k I vaguely remember you said it could use interfaces to abstract the difference out for the various versions, is that the same problem here ?

Reply:
> They need to decide if they want to work with the old API or the new API right? There should be some period of time in which both APIs are supported by the server.

Not all versioning is required to have a cut-over period.  An alpha to beta for example.  In addition, this keeps your tool constantly churning in reaction to server versions, making it much harder to divorce a release cycle of `kubectl` from `kube-apiserver`.

> How is this different from what clients built outside kubernetes/kubernetes need to do?

I still maintain that not exposing internal types is a mistake and we've seen poor reactions to it from every component trying to straddle api versions on multiple resources.  I'm not saying you can't or even shouldn't make this change, but you should be aware of the pain and risk you're incurring.  The HPA controller was one of the first to face this problem and it would not have been successful if it hadn't had the internal conversions to fall back on.  Hiding the internal types has shifted the debt and risk to every consumer.

Semi-related while I'm thinking about it, @DirectXMan12 @sttts we're probably ready to remove the conversions for the HPA controller now that we've cut over, right?

Reply:
> @deads2k I vaguely remember you said it could use interfaces to abstract the difference out for the various versions, is that the same problem here ?

That solution looks pretty awful.  See an example of just read-only access based on interfaces: https://github.com/openshift/origin/pull/16549

Reply:
> Not all versioning is required to have a cut-over period. An alpha to beta for example. 

Commands that use alpha apis should be under the `kubectl alpha` subcommand.  We should not be trying to preserve version skew for these items in the client.

> I'm not saying you can't or even shouldn't make this change, but you should be aware of the pain and risk you're incurring.

Appreciate the warning.  Sounds like we are in a broken state right now.  Do we have any options besides the following?

- build clis in the core repo (clearly not an option that will scale for the ecosystem, and doing so prevents us from releasing fixes to kubectl independently from the server)
- clis vendor the unversioned types from kubernetes/kubernetes (my understanding is that this is unsupported and comes with its own challenges)
- clis use external clients and suffer from the issues you described

Is this something apimachinery is planning on resolving in the near term?

Reply:
> Commands that use alpha apis should be under the kubectl alpha subcommand. We should not be trying to preserve version skew for these items in the client.

What about `kubectl describe`?  How do we deal with alpha versions there?  Do we have `kubectl alpha describe` too?

> Semi-related while I'm thinking about it, @DirectXMan12 @sttts we're probably ready to remove the conversions for the HPA controller now that we've cut over, right?

I'll file an issue.

Reply:
> What about kubectl describe? How do we deal with alpha versions there? Do we have kubectl alpha describe too?

That is a special case.  If you are describing an alpha API, expect it to break between releases.